README
NAME
    CatalystX::SimpleAPI - Simple API support for Catalyst apps

SYNOPSIS
        package ServiceApp::Controller::API;

        use Moose;
        use namespace::autoclean;
        BEGIN { extends 'Catalyst::Controller' }

        with 'CatalystX::Controller::SimpleAPI';

        __PACKAGE__->config(
            authkeys => {
                'AE281S228D4' => {
                    ip_check => '^10\.0\.0\.[0-9]+$',
                    valid_applications => 'myapp',
                },
            },
        );

        sub auto : Private {
            my ( $self, $c ) = @_;

            # will return false if the api did not pass authorization.
            return $self->prepare_api_request($c);
        }

        sub foo : Local {
            my ( $self, $c ) = @_;
            $c->stash(
                api_response => {
                    processed => 1,
                    status => 'success',
                    data => {
                        results => $c->req->params,
                    }
                },
            );
        }

        1;

        ...

        package MyApp::Model::ServiceApp; 

        use Moose;
        use namespace::autoclean;

        extends 'Catalyst::Model';

        with 'SimpleAPI::Agent';

        __PACKAGE__->config(
            api_key => 'AE281S228D4',
            application_id => 'myapp',
            api_base_url => 'http://localhost:5000/'
        );

        1;

        ...

        $c->model('ServiceApp')->request('/api/foo', { value => 10 });

DESCRIPTION
    It provides a simple API support - currently only JSON based - for
    Catalyst applications.

USAGE
    In your actions, all your API arguments will have already been placed in
    "$c->stash->{'api_params'}". Your API response will be placed in
    "$c->stash->{'api_response'}". The "api_response" hash should have the
    following structure.

        $c->stash->{'api_response'} = {
            # true / false value indicating the request was processed.
            processed => 0,

            # 'success' or 'failed' indicating whether the request was processed
            # successfully.
            status => 'failed',
    
            # hashref containing the results of the api action
            data => {},

            # error messages in the form of field => 'message' in the case of an error
            errors => {
                # 'general' should always be present to describe the overall error
                general => 'failed validation',

                # per-parameter error messaging if appropriate.
                ingredient => 'Ingredient is invalid.',
            },
        };

    If you process your request, you should set "processed" to 1 even if the
    request was did not have a successful result. The "processed" value is
    used to indicate that the parameters were accepted and the action
    requested was attempted. The "status" value is used to indicate whether
    the requested action accomplished what was requested. The basic rule of
    thumb here is that "processed" should only be set to 0 if the action
    could not be started for some reason (such as auth failure or other
    exceptional condition.) Note that if "processed" is 0, "status" will
    ALWAYS be failed.

    In your subclass, you will need to call the
    "$self->prepare_api_request($c)" method to initialize the API request.
    This is usually done in the "auto" action or the root of the action
    chain.

SEE ALSO
    Catalyst Advent Calendar 2009
    <http://www.catalystframework.org/calendar/2009/21>

AUTHOR
    Jay Kuri (jayk) "<jayk@cpan.org>".

CONTRIBUTORS
    Wallace Reis (wreis) "<wreis@cpan.org>".

SPONSORSHIP
    Development sponsored by Ionzero LLC <http://www.ionzero.com/>.

COPYRIGHT & LICENSE
    Copyright (C) 2010 Jay Kuri and the above contributors.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for CatalystX::SimpleAPI

0.04    1 Mar 2011
    - Fix Content-Type setting

0.03    14 Jan 2011
    - Move utf8 test to author mode

0.02    13 Jan 2011
    - Add JSON serialization of request params

0.01    18 Dec 2010
    - First release


t/lib/Simple.pm
package Simple;

use Moose;
use namespace::autoclean;
use Catalyst::Runtime 5.80;
use CatalystX::RoleApplicator;

extends 'Catalyst';

our $VERSION = '0.01';

__PACKAGE__->config(name => 'Simple');

__PACKAGE__->apply_request_class_roles(qw/
    Catalyst::TraitFor::Request::DecodedParams::JSON
/);

__PACKAGE__->setup;

1;


t/lib/Simple/Controller/API.pm
package Simple::Controller::API;

use Moose;
use namespace::autoclean;

BEGIN { extends 'Catalyst::Controller' }

with 'CatalystX::Controller::SimpleAPI';

__PACKAGE__->config(
    authkeys => {
        'AE281S228D4' => {
            ip_check => '^127',
            valid_applications => 'simple',
        },
    },
);

sub auto : Private {
    my ( $self, $c ) = @_;
    return $self->prepare_api_request($c);
}

sub foo : Local {
    my ( $self, $c ) = @_;
    $c->stash(
        api_response => {
            processed => 1,
            status => 'success',
            data => $c->req->decoded_params->{'value'},
        },
    );
}

sub return_error : Local {
    my ( $self, $c ) = @_;
    $c->stash(
        api_response => {
            processed => 1,
            status => 'fail',
            errors => {
                general => ['Error in API'],
            },
        },
    );
}

__PACKAGE__->meta->make_immutable;

1;


lib/SimpleAPI/Agent.pm
package SimpleAPI::Agent;

use Moose::Role;
use namespace::autoclean;
use Moose::Util::TypeConstraints;
use LWP::UserAgent;
use JSON::Any;
use URI;

our $VERSION = '0.02';

has 'user_agent' => ( 
    isa => duck_type([qw/get post/]), is => 'ro',
    lazy => 1, builder => '_build_user_agent',
);

sub _build_user_agent { return LWP::UserAgent->new }
                
has 'api_key' => ( isa => 'Str', is => 'ro', required => 1 );

has 'application_id' => ( isa => 'Str', is => 'ro', required => 1 );

has 'api_base_url' => ( isa => 'Str', is => 'ro', required => 1 );

has 'json_decoder' => (
    isa => duck_type([qw/jsonToObj/]), is => 'ro',
    lazy => 1, builder => '_build_json_decoder',
);

sub _build_json_decoder { return JSON::Any->new }

has 'json_encoder' => (
    isa => duck_type([qw/encode/]), is => 'ro',
    lazy => 1, builder => '_build_json_encoder',
);

sub _build_json_encoder { return JSON::Any->new }

sub _serialize_params {
    my ( $self, $param ) = @_;
    $param ||= {};
    my $json_encoder = $self->json_encoder;
    foreach my $key ( keys %$param ) {
        my $data = $param->{$key};
        next unless ref $data;
        $param->{$key} = $json_encoder->encode($data);
    }
    return $param;
}

sub request {
    my ( $self, $path, $data, $method ) = @_;
    confess 'The params needs to be in a hashref'
        if defined $data && ref($data) ne 'HASH';
    $data ||= {};
    $method = uc($method || 'POST');
    $data->{'application'} = $self->application_id;
    if (!defined($data->{'authkey'})) {
        $data->{'authkey'} = $self->api_key;
    }
    my $base_url = $self->api_base_url;
    $base_url =~ s{/$}{}g;
    $path =~ s{^/}{}g;
    my $uri = URI->new(join(q{/}, $base_url, $path));

    return $self->handle_response(
        $self->do_request($method, $uri, $data)
    );
}

sub do_request {
    my ( $self, $method, $uri, $data ) = @_;
    confess 'The params needs to be in a hashref'
        if defined $data && ref($data) ne 'HASH';
    $data = $self->_serialize_params({ %$data });
    my $req_method = q{_} . lc $method . '_request';
    if ( $self->can($req_method) ) {
        return $self->$req_method($uri, $data);
    }
    else {
        confess "$method not supported";
    }
}

sub _get_request {
    my ( $self, $uri, $data) = @_; 
    $uri->query_form($data);
    return $self->user_agent->get($uri);
}

sub _post_request {
    my ( $self, $uri, $data ) = @_; 
    return $self->user_agent->post($uri, $data);
}

sub handle_response {
    my ( $self, $response ) = @_;
    if ( $response->is_success ) {
        my $response_data = $self->json_decoder->jsonToObj($response->content);
        if ( (exists($response_data->{'processed'})
                && $response_data->{'processed'} == 0)
            || (exists($response_data->{'status'})
                && $response_data->{'status'} ne 'success') 
        ) {
            confess $response_data->{'errors'};
        } else {
            return $response_data->{'data'};
        }
    } else {
        confess 'Request to SimpleAPI failed: ' . $response->status_line;
    }
}

1;

__END__

=head1 NAME

SimpleAPI::Agent - Role for custom user agents

=head1 SYNOPSIS

    package MyApp::Model::ServiceApp;

    use Moose;
    use namespace::autoclean;

    extends 'Catalyst::Model';

    with 'SimpleAPI::Agent';

    __PACKAGE__->config(
        api_key => 'AE281S228D4',
        application_id => 'myapp',
        api_base_url => 'http://localhost:5000/'
    );

    1;

=head1 DESCRIPTION

A role for building custom user agents which provides a C<request> method to
interface with L<CatalystX::SimpleAPI> apps.

=head1 ATTRIBUTES

=over

=item api_key

String used as authorization token.

=item application_id

Application's identification which can be a number, UUID, name, etc.

=item api_base_url

Base URL to send request to. That is concatenated with the path supplied as arg
on the call to C<request>.

=back

=head1 METHODS

=over

=item C< request( $relative_path, $params?, $http_method? ) >

=item C< do_request( $http_method, $uri_obj, $params? ) >

=item C< handle_response( $http_response_obj ) >

=back

=head1 AUTHOR & LICENSE

See L<CatalystX::SimpleAPI>.

=cut


lib/CatalystX/SimpleAPI.pm
package CatalystX::SimpleAPI;

our $VERSION = '0.04';

1;

__END__

=head1 NAME

CatalystX::SimpleAPI - Simple API support for Catalyst apps

=head1 SYNOPSIS

    package ServiceApp::Controller::API;

    use Moose;
    use namespace::autoclean;
    BEGIN { extends 'Catalyst::Controller' }

    with 'CatalystX::Controller::SimpleAPI';

    __PACKAGE__->config(
        authkeys => {
            'AE281S228D4' => {
                ip_check => '^10\.0\.0\.[0-9]+$',
                valid_applications => 'myapp',
            },
        },
    );

    sub auto : Private {
        my ( $self, $c ) = @_;

        # will return false if the api did not pass authorization.
        return $self->prepare_api_request($c);
    }

    sub foo : Local {
        my ( $self, $c ) = @_;
        $c->stash(
            api_response => {
                processed => 1,
                status => 'success',
                data => {
                    results => $c->req->params,
                }
            },
        );
    }

    1;

    ...

    package MyApp::Model::ServiceApp; 

    use Moose;
    use namespace::autoclean;

    extends 'Catalyst::Model';

    with 'SimpleAPI::Agent';

    __PACKAGE__->config(
        api_key => 'AE281S228D4',
        application_id => 'myapp',
        api_base_url => 'http://localhost:5000/'
    );

    1;

    ...

    $c->model('ServiceApp')->request('/api/foo', { value => 10 });

=head1 DESCRIPTION

It provides a simple API support - currently only JSON based - for Catalyst
applications.

=head1 USAGE

In your actions, all your API arguments will have already been placed in
C<< $c->stash->{'api_params'} >>. Your API response will be placed in
C<< $c->stash->{'api_response'} >>. The C<api_response> hash should 
have the following structure.

    $c->stash->{'api_response'} = {
        # true / false value indicating the request was processed.
        processed => 0,

        # 'success' or 'failed' indicating whether the request was processed
        # successfully.
        status => 'failed',
    
        # hashref containing the results of the api action
        data => {},

        # error messages in the form of field => 'message' in the case of an error
        errors => {
            # 'general' should always be present to describe the overall error
            general => 'failed validation',

            # per-parameter error messaging if appropriate.
            ingredient => 'Ingredient is invalid.',
        },
    };

If you process your request, you should set C<processed> to B<1> even if the
request was did not have a successful result.  The C<processed> value is used to
indicate that the parameters were accepted and the action requested was attempted.
The C<status> value is used to indicate whether the requested action accomplished
what was requested.  The basic rule of thumb here is that C<processed> should only
be set to B<0> if the action could not be started for some reason (such as auth
failure or other exceptional condition.)  Note that if C<processed> is B<0>, C<status>
will ALWAYS be B<failed>.  

In your subclass, you will need to call the C<< $self->prepare_api_request($c) >> 
method to initialize the API request. This is usually done in the C<auto>
action or the root of the action chain.

=head1 SEE ALSO

L<Catalyst Advent Calendar 2009|http://www.catalystframework.org/calendar/2009/21>

=head1 AUTHOR

Jay Kuri (jayk) C<< <jayk@cpan.org> >>.

=head1 CONTRIBUTORS

Wallace Reis (wreis) C<< <wreis@cpan.org> >>.

=head1 SPONSORSHIP

Development sponsored by Ionzero LLC L<http://www.ionzero.com/>.

=head1 COPYRIGHT & LICENSE

Copyright (C) 2010 Jay Kuri and the above contributors.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
